Extend GroundPrimitive shadow volume based on screen space error

hpinkos opened the pull request:
Fixes #4326
Fixes #4161

This extends the bottom of the shadow volume below the terrain based on the screen space error.  The shadow volume is updated in the ShadowVolume vertex shader.

- Ground primitives geometry types compute a `extrudeDirection` attribute.  The `extudeDirection` is a normalized vector pointing in the direction in which the position should be moved in the vertex shader.  
  - Positions at the top of the shadow volume will have `extrudeDirection === Cartesian3.ZERO` because only the bottom positions need to be moved
- Added `czm_geometricToleranceOverDistance ` to `AutomaticUniforms`
- Added `pixelSizePerMeter` and `maximumScreenSpaceError` to `UniformState` 
  - `scene.globe.maximumScreenSpaceError` get piped through to `UniformState` via the `frameState`.  Not sure if that's the best way to do this.
- Added `u_globeMinimumAltitude` to `GroundPrimitive.uniforms`.  
  - I set this to `33000.0`, which is a magic number I found with testing.  I'm not sure what this value should actually be. 
  - I made this a positive number so I don't have to `abs` it later when i find the `min` between this value and the value computed using the screen space error.

TODO
- [x] Fix debugShowShadowVolume so it displays the shadow volume after the positions have been updated in the shader
- [x] octencode `extrudeDirection` attribute
- [x] Update CHANGES.md

pjcozzi commented:
> Ground primitives geometry types compute a extrudeDirection attribute.

Let's octencode this.  See [AttributeCompression.js](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/AttributeCompression.js).

pjcozzi commented:
> I set this to 33000.0, which is a magic number I found with testing. I'm not sure what this value should actually be.

This should come from the terrain provider, but we can hard code it to the right min for Earth for now.  @bagnell do you have it handy from STK?

pjcozzi commented:
> Positions at the top of the shadow volume will have extrudeDirection === Cartesian3.ZERO because only the bottom positions need to be moved

@lilleyse @bagnell do you guys have any ideas on how to avoid this wasted memory without introducing another draw call per volume?

pjcozzi commented:
Update CHANGES.md.

pjcozzi commented:
Perhaps there is a winding order issue:

![image](https://cloud.githubusercontent.com/assets/782098/21196284/12f204e2-c205-11e6-959d-afd9cdef3592.png)

Click "Create large polygons"

http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/ground-primitive-sse/Apps/Sandcastle/index.html?src=development/Ground%20Primitive.html&label=Showcases

pjcozzi commented:
>> Positions at the top of the shadow volume will have extrudeDirection === Cartesian3.ZERO because only the bottom positions need to be moved

> @lilleyse @bagnell do you guys have any ideas on how to avoid this wasted memory without introducing another draw call per volume?

Related idea: at one point, I just tried to normalize the WGS84 position, to avoid storing an extra attribute, and it was very noticeably wrong at places like Mount Saint Helens.  However, perhaps there is another way to compute a better approximation, but still not perfect.

pjcozzi commented:
Also with that same example, the corridor and circle disappear when zoomed out.

![bv](https://cloud.githubusercontent.com/assets/782098/21196499/c96d303e-c205-11e6-94b5-3206810d583d.gif)

pjcozzi commented:
@hpinkos did you try using `GroundPrimitive.debugShowShadowVolume` to verify that the volumes are as low as they need to be (so it renders correctly), but not much lower (to avoid fillrate)?

pjcozzi commented on Source/Renderer/AutomaticUniforms.js:
```diff
@@ -1476,6 +1476,20 @@ define([
+         * @alias czm_geometricToleranceOverDistance
+         * @glslUniform
+         */
+        czm_geometricToleranceOverDistance : new AutomaticUniform({
```
`Distance` -> `Meter`

pjcozzi commented on Source/Renderer/UniformState.js:
```diff
@@ -935,6 +961,14 @@ define([
+
+        var fov = camera.frustum.fov;
+        var viewport = this._viewport;
+        if (viewport.width > viewport.height * camera.frustum.aspectRatio) {
```
How does this check work?

hpinkos replied:
This was just a straight port.  Did I misinterpreted something? Here's the original source:
``` c
    if ((double)mainViewport[3] > ((double)mainViewport[2] * p3dScene->pixelAspect))
    {
        pixelSizePerDistance = tan(0.5 * p3dScene->view.m_pCamera->FieldOfView()) * 2.0 / (double)mainViewport[3];
    }
    else
    {
        pixelSizePerDistance = tan(0.5 * p3dScene->view.m_pCamera->FieldOfView()) * 2.0 / (double)mainViewport[2];
    }
```

pjcozzi replied:
Are you sure `mainViewport[3]` is `width`?  In Cesium, the array would be `[x, y, width, height]`.

Are you sure `pixelAspect` ratio is `width / height` like `aspectRatio`?

Perhaps ask Greg to double check these.

Right now, I think the `if` statements reads like "if (width > (height * (width / height))", which is "if (width > width)"

pjcozzi commented on Source/Shaders/ShadowVolumeVS.glsl:
```diff
@@ -17,7 +20,11 @@ vec4 depthClampFarPlane(vec4 vertexInClipCoordinates)
+    float delta = min(u_globeMinimumAltitude, czm_geometricToleranceOverDistance * length(position.xyz));
+
+    //extrudeDirection is zero for the top layer
+    position = position + vec4(extrudeDirection.xyz * delta, 0);
```
Use `0.0`, mixing integers and floats didn't use to compile on some WebGL implementations.

pjcozzi replied:
`extrudeDirection` is a `vec3` so we can remove the `.xyz `.

pjcozzi commented:
What do you think about cleaning up the geometry types by adding adding a temporary `BOTTOM` (or whatever) attribute then running the `Geometry` through a generic function that removes that attribute and adds the extruded normal?  It would come at a potential small performance and GC cost but could be worth it.

@bagnell what do you think about this for vector tiles?  Seems like we don't need to store the extruded direction in the payload.  We could even separate out the bottom cap vertices so the `BOTTOM` attribute is implicit over the wire.

I feel like we should do some fill rate tests before getting into all this.  This should make a huge difference for horizon and 45 degree views depending on how well early-z does (even with early-z the rasterization load is very high).

hpinkos commented:
Here is a visualization of the shadow volume.  The blue is the old volume and the red is the new one.  It just extends a little bit below what we were doing previously

large polygon:
![image](https://cloud.githubusercontent.com/assets/3451886/21328498/41cbb320-c602-11e6-9d0a-23511245bd32.png)

small rectangle:
![image](https://cloud.githubusercontent.com/assets/3451886/21328500/44f620f8-c602-11e6-8019-ce82333745b0.png)

hpinkos commented:
@pjcozzi I believe I addressed your comments and fixed the problems you found with some of the geometries.

I haven't done the octencodeing yet because I'm waiting to see if we want to change the `extrudeDirection` attribute

pjcozzi commented:
> Fix debugShowShadowVolume so it displays the shadow volume after the positions have been updated in the shader

Added in 2d50df1ef5a9c878e5fd75020fc58c4bcc8c9191.

This restriction was removed:

>Must be <code>true</code> on creation for the volumes to be created before the geometry is released or releaseGeometryInstances must be <code>false</code>

But the polygon is no longer drawn when the volume is drawn.

pjcozzi commented:
> I haven't done the octencodeing yet because I'm waiting to see if we want to change the extrudeDirection attribute

Let's just move forward so we don't stall this.  We can optimize later if we come up with a good idea.

pjcozzi commented on Source/Renderer/AutomaticUniforms.js:
```diff
@@ -1476,6 +1476,20 @@ define([
+            size: 1,
+            datatype: WebGLConstants.FLOAT,
+            getValue: function(uniformState) {
+                return uniformState.pixelSizePerMeter * uniformState.maximumScreenSpaceError;
```
This is constant per frame, right?  Precompute and store in `UniformState`.

pjcozzi commented on Source/Renderer/UniformState.js:
```diff
@@ -768,6 +770,30 @@ define([
+         * @type {Number}
+         * @default 2
+         */
+        maximumScreenSpaceError: {
```
These two getters may not be needed after the about change.

pjcozzi commented:
Just those comments.

hpinkos commented:
@pjcozzi this is ready for another look.  I fix the issues you pointed out and I added octencode for the new attribute.  And I updated `CHANGES.md`

bagnell commented:
Oct-encoding is fine for lighting because the error isn't noticeable but it may be noticeable for geometry extruded with these normals. Have you seen this at all?

hpinkos commented:
I didn't see any noticeable errors, but I'll look more closely to make sure

pjcozzi commented:
> Oct-encoding is fine for lighting because the error isn't noticeable but it may be noticeable for geometry extruded with these normals

How many bits are we using?  It should be OK if we use enough....more than lighting.  For example, normals for rotation of instances in i3dm use 32: https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/TileFormats/Instanced3DModel#oct-encoded-normal-vectors

The paper has exact error bounds.

pjcozzi commented:
Bump:

>> Oct-encoding is fine for lighting because the error isn't noticeable but it may be noticeable for geometry extruded with these normals
> How many bits are we using? It should be OK if we use enough....more than lighting. For example, normals for rotation of instances in i3dm use 32: https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/TileFormats/Instanced3DModel#oct-encoded-normal-vectors

> The paper has exact error bounds.

hpinkos commented:
Thanks @bagnell, @pjcozzi, I'll update it to use 32

hpinkos commented:
@pjcozzi @bagnell ready

pjcozzi commented on Source/Shaders/Builtin/Functions/octDecode.glsl:
```diff
@@ -1,3 +1,29 @@
+  * @param {float} range The maximum value of the SNORM range. The encoded vector is stored in log2(rangeMax+1) bits.
+  * @returns {vec3} The decoded and normalized vector
+  */
+  vec3 czm_octDecode(vec2 encoded, float range)
```
I'll be pleasantly surprised if JSDoc handles function overloads, but it's not an issue now.

pjcozzi commented:
Do you get these artifacts?  If so, do you also get them without normal compression?

![ok2](https://cloud.githubusercontent.com/assets/782098/21702272/954d1010-d378-11e6-967d-f3875f9a7546.gif)

If these are due to normal compression, is there anything we can do about it?  Are we sure the code is right?  Is there a more precise (48 bits?) version from the paper we can try?

hpinkos commented:
@pjcozzi no I don't see that on my machine.  That's happening after the 32P update?

pjcozzi commented:
I don't get this on Windows.  Will test Mac/Intel now, hopefully the artifact is in master.

pjcozzi commented:
Good news, it is in master.  See #4820

pjcozzi commented:
OK, this is good!

Please reply to all the forum threads.  People will be thrilled!
